### PR TITLE
SAT: backward compatibility - check that cursor fields were not changed

### LINKED ISSUE #15439: SAT: test `STATE` backward compatibility (syntactic checks)
## Tell us about the problem you're trying to solve
When upgrading a connector, one of the most important things is to ensure that your sync progress will not be reset.
This new connector version should still be able to read any previously created STATE file.

## What can make a `STATE` not backward compatible:
If the developer changes the cursor field or the logic to update the cursor value, this can break the post-upgrade sync in the following manner:
1. The current connector version loads a state generated with the previous connector version
2. The new cursor field is not present in the previous connector state
3. Past data, potentially already synced to a destination, get backfilled, leading to duplicates on the destination and unexpected load on the source.

Because `STATE` objects are freeform objects, any key it contains might be used to alter a sync logic, so for safety, we can consider changes on all keys potentially breaking changes.  

## Potential **syntactic** tests: 
1. **Check that the cursor paths did not change:**
   * Run `discover` on both versions
   * Assert that the cursor field on each stream did not change

Removed after grooming
~~2. **Check that the previous STATE keys are a subset of the current STATE keys**:
    *  Run `read` on both connector versions, retrieve STATE messages
    *  `assert set(previous_state.keys()).issubset(set(current_state.keys()))`
    * Risk: performing `read` with the same config on both versions asserts that backward compatibility tests for spec passed.~~

## Acceptance criteria:

* Changing a cursor field on a stream makes the `incremental` test suite fail on SAT
~~* Removing keys from a state object makes the `incremental` test suite fail on SAT 
* The backward compatibility test can be bypassed by editing `acceptance-test-config.yml`~~

### PR DESCRIPTION
**Pull request by alafanechere**
## What
Closes #15439

We want to make backward compatibility test fail if a `default_cursor_field` value was changed in a stream of the  connector catalog.

## How
* Retrieve previous and current catalog with a `discover` run (previously implemented in #15486) 
* Refacto:
    * `BaseDiffChecker` and its child classes now perform the diff computation within the class
    * Multiple diff can be computed on different dict, so different diff can be used in the `check_*` function. This is useful because checking `default_cursor_value` change does not happen a the same diff as `json_schema` change.
    * Remove the diff computation from `test_core.py`
* Add `CatalogDiffChecker.check_if_cursor_field_was_changed` with alongside test cases in `test_backward_compatibility.FAILING_CATALOG_TRANSITIONS`
 * Bump SAT to `0.2.0` as I consider this backward compatibility effort (on syntactic validation) done.

## Recommended reading order
1. [test_backward_compatibility.py](https://github.com/airbytehq/airbyte/pull/15520/files#diff-c420c54a7033f7e3d14e5c459739a8670ce8369d57e9ceb44a8725d738c2b7a6)
2. [backward_compatibility.py](https://github.com/airbytehq/airbyte/pull/15520/files#diff-cb3384a3065d3872dac36499411e6ce74860f317bcc9a7d902596954ed584743)
3. [test_core.py](https://github.com/airbytehq/airbyte/pull/15520/files#diff-404bcf3ff97587546648373b392bee96c021274dfd8508a07c6b7ec83780a3a8)

## 🚨 User Impact 🚨
Developers changing the `default_cursor_field` to another value will face SAT failure.

### CODE REVIEW

File: airbyte-integrations/bases/source-acceptance-test/unit_tests/test_backward_compatibility.py [hunk 1]
```diff
@@ -915,6 +915,7 @@ def test_spec_backward_compatibility(previous_connector_spec, actual_connector_s
 ]
 
 
+@pytest.mark.slow
```
Comment (sherifnada) [inline airbyte-integrations/bases/source-acceptance-test/unit_tests/test_backward_compatibility.py:918]: what's the value of this mark?

**Review by sherifnada — APPROVED** [review]
lgtm!

File: airbyte-integrations/bases/source-acceptance-test/unit_tests/test_backward_compatibility.py (hunk 1 above)
Comment (alafanechere) [reply airbyte-integrations/bases/source-acceptance-test/unit_tests/test_backward_compatibility.py:918]: The decorated test is slow to run,  SAT developers can decide to bypass it while iterating on their change by running `python -m pytest unit_tests -m "not slow"`

Comment (sherifnada) [reply airbyte-integrations/bases/source-acceptance-test/unit_tests/test_backward_compatibility.py:918]: is that a known pythonic standard? might be helpful to include in the docs

Comment (alafanechere) [reply airbyte-integrations/bases/source-acceptance-test/unit_tests/test_backward_compatibility.py:918]: I thought it was a standard as it's the first example in pytest doc about marks:
https://docs.pytest.org/en/7.1.x/how-to/mark.html#registering-marks

Comment (sherifnada) [reply airbyte-integrations/bases/source-acceptance-test/unit_tests/test_backward_compatibility.py:918]: I am by no means the yardstick of pythonic knowledge, so take this with a grain of salt, but if there's a chance users may not know about it, it's helpful to add to the docs

Comment (alafanechere) [reply airbyte-integrations/bases/source-acceptance-test/unit_tests/test_backward_compatibility.py:918]: 👍 Sure, if you ask, it must be documented.

**Comment by alafanechere** [comment]
/publish connector=bases/source-acceptance-test auto-bump-version=false

**Pull request merged**